workers/unity/Packages/com.improbable.gdk.playerlifecycle/Systems/PlayerHeartbeat/HandlePlayerHeartbeatResponseSystem.cs

```diff
@@ -4,87 +4,78 @@
                 foreach (var response in responses)
                 {
-                    if (response.StatusCode != StatusCode.Success)
+                    if (response.StatusCode == StatusCode.Success)
                     {
-                        var failedHeartbeatsComponent = EntityManager.HasComponent<HeartbeatData>(entity)
-                            ? EntityManager.GetComponentData<HeartbeatData>(entity)
-                            : new HeartbeatData();
-                        failedHeartbeatsComponent.NumFailedHeartbeats++;
-                        if (EntityManager.HasComponent<HeartbeatData>(entity))
-                        {
-                            PostUpdateCommands.SetComponent(entity, failedHeartbeatsComponent);
-                        }
-                        else
-                        {
-                            PostUpdateCommands.AddComponent(entity, failedHeartbeatsComponent);
-                        }
+                        stillAlive = true;
+                        break;
+                    }
+                }
 
-                        Debug.LogFormat(Messages.FailedHeartbeat, entityId,
-                            failedHeartbeatsComponent.NumFailedHeartbeats,
-                            PlayerLifecycleConfig.MaxNumFailedPlayerHeartbeats);
+                if (stillAlive)
+                {
+                    heartbeatsComponent.NumFailedHeartbeats = 0;
+                }
+                else
+                {
+                    heartbeatsComponent.NumFailedHeartbeats++;
 
-                        if (failedHeartbeatsComponent.NumFailedHeartbeats >=
-                            PlayerLifecycleConfig.MaxNumFailedPlayerHeartbeats)
-                        {
-                            Debug.LogFormat(Messages.DeletingPlayer, entityId);
-                            entityDeleteSender.RequestsToSend.Add(WorldCommands.DeleteEntity.CreateRequest
-                            (
-                                entityId
-                            ));
-                            data.WorldCommandSenders[i] = entityDeleteSender;
-                        }
-                    }
-                    else
+                    if (heartbeatsComponent.NumFailedHeartbeats >=
+                        PlayerLifecycleConfig.MaxNumFailedPlayerHeartbeats)
                     {
-                        if (EntityManager.HasComponent<HeartbeatData>(entity))
-                        {
-                            PostUpdateCommands.RemoveComponent<HeartbeatData>(entity);
-                        }
+                        var entityDeleteSender = heartbeatGroup.WorldCommandSenders[i];
+                        entityDeleteSender.RequestsToSend.Add(WorldCommands.DeleteEntity.CreateRequest
+                        (
+                            heartbeatGroup.SpatialEntityIds[i].EntityId
+                        ));
+                        heartbeatGroup.WorldCommandSenders[i] = entityDeleteSender;
```

Comment:
are line s68 and 73  needed? we only use `entityDeleteSender` once, so we could just use it directly instead of assigning it and then re-assigning it back to `heartbeatGroup.WorldCommandSenders[i]`

Reply:
Because assigning it directly does not write back to the ComponentDataArray

Reply:
but do we need to assign it back when adding a request to the list?

Reply:
no, because the list Doesn't care. But this will allow us to use the change filters.